lib/features/modeling/cmd/ToggleShapeCollapseHandler.js

```diff
@@ -55,26 +60,46 @@ ToggleShapeCollapseHandler.prototype.revert = function(context) {
  *
  * @return {Object}
  */
-function getElementsVisibility(elements) {
+function getElementsVisibilityRecursive(elements) {
 
   var result = {};
 
-  elements.forEach(function(e) {
-    result[e.id] = e.hidden;
+  forEach(elements, function(element) {
+    result[element.id] = element.hidden;
+
+    if (element.children) {
+      result = assign({}, result, getElementsVisibilityRecursive(element.children));
+    }
   });
 
   return result;
 }
 
 
-function setHidden(elements, newHidden) {
-  elements.forEach(function(element) {
+function setHiddenRecursive(elements, newHidden) {
+  var result = [];
+  forEach(elements, function(element) {
     element.hidden = newHidden;
+    result = result.concat(element);
+
+    if (element.children) {
+      result = result.concat(setHiddenRecursive(element.children, newHidden));
+    }
   });
+
+  return result;
 }
 
-function restoreVisibility(elements, lastState) {
-  elements.forEach(function(e) {
-    e.hidden = lastState[e.id];
+function restoreVisibilityRecursive(elements, lastState) {
```

Comment:
This does not properly handle expanding when a nested child element is collapsed, see the failing test case I've created: https://github.com/bpmn-io/diagram-js/pull/421/commits/2a9da7cffd92a0d1eb2c225b45d6c4ed7b432d6e

Reply:
Ok I commited a fix about this. The restoreVisibilityRecursive actually works fine (undoing stuff), the problem was with the setHiddenRecursive method instead.